File: go/src/utils/vlq.go
package utils

import (
	. "math/big"
	"strings"
)

var BASE64_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"

var VLQ_BASE_SHIFT = uint(5)
var VLQ_BASE = NewInt(0b100000) // 0b100000
var VLQ_CONTINUATION_BIT = uint64(0b100000)
var VLQ_BASE_MASK = NewInt(0b11111)

var ONE = NewInt(1)

func Encode(ints []Int) string {
	out := ""

	for _, x := range ints {
		x = toVLQSigned(x)
		for _, b := range vlqEncode(x) {
			out += base64Encode(b)
		}
	}

	return out
}

func Decode(data string) []Int {
	out := []Int{}
	vlqs := [][]byte{}
	vlq := []byte{}

	for _, x := range data {
		sextet := base64Decode(string(x))
		vlq = append([]byte{byte(sextet)}, vlq...)
		if (uint64(sextet) & VLQ_CONTINUATION_BIT) == 0 {
			vlqs = append(vlqs, vlq)
			vlq = []byte{}
		}
	}

	for _, vlq := range vlqs {
		x := vlqDecode(vlq)
		x = fromVLQSigned(x)
		out = append(out, x)
	}

	return out
}

func vlqDecode(vlq []byte) Int {
	x := NewInt(0)
	for _, sextet := range vlq {
		x.Lsh(x, VLQ_BASE_SHIFT)
		s := NewInt(int64(sextet))
		x.Or(x, s.And(s, VLQ_BASE_MASK))
	}
	return *x
}

func fromVLQSigned(x Int) Int {
	isNeg := NewInt(0).And(&x, ONE).Sign() == 1
	x.Rsh(&x, 1)
	if isNeg {
		x.Neg(&x)
	}
	return x
}

func toVLQSigned(x Int) Int {
	isNeg := x.Sign() == -1
	x.Abs(&x)
	x.Lsh(&x, 1)

	if isNeg {
		x.Add(&x, ONE)
	}
	return x
}

func vlqEncode(x Int) []byte {
	if x.Sign() == 0 {
		return []byte{0}
	}

	sextets := []byte{}
	for x.Sign() > 0 {
		sextet := NewInt(0).And(&x, VLQ_BASE_MASK).Uint64()
		x.Rsh(&x, 5)
		if x.Sign() > 0 {
			sextet = sextet | VLQ_CONTINUATION_BIT
		}
		sextets = append(sextets, byte(sextet))
	}
	return sextets
}

func base64Decode(char string) int {
	return strings.Index(BASE64_ALPHABET, char)
}

func base64Encode(sextet byte) string {
	return string(BASE64_ALPHABET[sextet])
}

File: deno/src/vlq.ts
const BASE64_ALPHABET =
  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

const char_to_integer = new Map<string, number>();

BASE64_ALPHABET
	.split('')
	.forEach(function (char, i) {
		char_to_integer.set(char, i);
	});

const VLQ_BASE_SHIFT = 5n;
const VLQ_BASE = 1n << VLQ_BASE_SHIFT; // 0b100000
const VLQ_CONTINUATION_BIT = Number(VLQ_BASE); // 0b100000
const VLQ_BASE_MASK = VLQ_BASE - 1n; // 0b11111n

/**
 * Returns the base 64 VLQ encoded values.
 */
export function encode(integers: bigint[]): string {
  return integers
    .map(vlqEncode)
    .map(base64Encode)
    .join("");
}

/**
 * Decodes the next base 64 VLQ values from the given string
 */
export function decode(base64Vlqs: string): bigint[] {
  return splitVlqs(base64Decode(base64Vlqs)).map(vlqDecode);
}

/**
 * Converts from a two-complement value to a value where the sign bit is
 * placed in the least significant bit.  For example, as decimals:
 *   1 becomes 2 (10 binary), -1 becomes 3 (11 binary)
 *   2 becomes 4 (100 binary), -2 becomes 5 (101 binary)
 */
function toVLQSigned(aValue: bigint): bigint {
  return aValue < 0n
    ? ((-aValue) << 1n) + 1n
    : (aValue << 1n) + 0n;
}

/**
 * Converts to a two-complement value from a value where the sign bit is
 * placed in the least significant bit.  For example, as decimals:
 *   2 (10 binary) becomes 1, 3 (11 binary) becomes -1
 *   4 (100 binary) becomes 2, 5 (101 binary) becomes -2
 */
function fromVLQSigned(aValue: bigint): bigint {
  const isNegative = (aValue & 1n) === 1n;
  const shifted = aValue >> 1n;
  return isNegative
    ? -shifted
    : shifted;
}

/**
 * Returns the VLQ encoded value.
 */
function vlqEncode(x: bigint): number[] {
  x = toVLQSigned(x);
  const sextets = [];
  while (x > 0) {
    let sextet = 0;
    sextet = Number(x & VLQ_BASE_MASK);
    x >>= 5n;
    if (x > 0) {
      sextet |= VLQ_CONTINUATION_BIT;
    }
    sextets.push(sextet);
  }
  return sextets;
}

/**
 * Returns the base 64 VLQ encoded value.
 */
function base64Encode(sextets: number[]) {
  return sextets.map((sextet) => {
    return BASE64_ALPHABET[sextet];
  }).join("");
}

function base64Decode(base64Vlqs: string): number[] {
  return base64Vlqs.split("").map((c) => {
    const sextet = char_to_integer.get(c);
    if (sextet === undefined) {
      throw new Error(`${base64Vlqs} is not a valid base64 encoded VLQ`);
    }
    return sextet;
  });
}

function splitVlqs(sextets: number[]): number[][] {
  const splitVlqs: number[][] = [];
  let vlq: number[] = [];
  sextets.forEach((sextet) => {
    vlq.push(sextet);
    if ((sextet & VLQ_CONTINUATION_BIT) === 0) {
      splitVlqs.push(vlq);
      vlq = [];
    }
  });
  if (vlq.length > 0) {
    throw new Error("Malformed VLQ sequence: The last VLQ never ended.");
  }
  return splitVlqs;
}

function vlqDecode(sextets: number[]): bigint {
  const x = sextets.reverse().reduce((x, sextet) => {
    x <<= 5n;
    x |= BigInt(sextet) & VLQ_BASE_MASK;
    return x;
  }, 0n);
  return fromVLQSigned(x);
}
